CTP-4655 show in Feedback Tracker report and My Feedback block

opitz opened the pull request:
As there was no dedicated method to provide assessor/marker related figures I have created a new class and put it there.
So far methods to return all submissions and submissions in need of grading for the current user as assessor have been created.

The logic has been blatantly copied and adopted from the mod_coursework_object_renderer class leaving it there (so we have duplicate code for now but I assume that you may have refactored the renderers anyway?)

opitz committed:
CTP-4162: fixed var type for $moderation->stageidentifier

opitz committed:
CTP-4162: added Behat test (thanks, Leon)
addressed more dynamic properties

opitz committed:
CTP-4363: added Behat test (thanks, Leon)
addressed dynamic property issues

opitz committed:
CTP-4655: added static method to provide the number of needed markings per assessor
added static method to return submissions related to user as assessor

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+
+        if (!$coursework->has_multiple_markers()
+            && has_capability('mod/coursework:addagreedgrade', $coursework->get_context())
+            && !has_capability('mod/coursework:addinitialgrade', $context)) {
```
If we've populated `$context` surely this should be:

```
        if (!$coursework->has_multiple_markers()
                && has_capability('mod/coursework:addagreedgrade', $context)
                && !has_capability('mod/coursework:addinitialgrade', $context)) {
```

opitz replied:
indeed...

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+
+            foreach ($submissions as $sub) {
+                $submission = submission::find($sub);
+                if ( $submission->final_grade_agreed()) {
```
Rogue space <code>if <b>( $</b>submission->final_grade_agreed()) {</code>.

opitz replied:
removed

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+    public static function get_submissions_for_assessor($coursework): array {
+        global $USER;
+
+        $gradeblesub = [];
```
Do we mean <code>$grade<b>a</b>blesub</code>? Maybe call this something like `$cangrade` instead?

opitz replied:
Ha! I simply copied the existing code from here (https://github.com/ucl-isd/moodle-mod_coursework/blob/2e55cb46556b5973856b8fc6139f88909042acb2/renderers/object_renderer.php#L1790) and wrapped it into a new static method without paying much attention to detail... now renamed to "$assessorsubmissions"

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+
+        } else if (is_siteadmin($USER) ||
+            !$coursework->allocation_enabled() ||
+            has_any_capability(['mod/coursework:administergrades'], $context)) {
```
Only checking one capability so presumably this should be:

```
        } else if (is_siteadmin($USER) ||
                !$coursework->allocation_enabled() ||
                has_capability('mod/coursework:administergrades', $context)) {
```

opitz replied:
refactored

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+            foreach ($submissions as $sub) {
+                $submission = submission::find($sub);
+                if ($coursework->assessor_has_any_allocation_for_student($submission->reload()->get_allocatable()) ||
+                    (has_capability('mod/coursework:addagreedgrade', $context))
```
Redundant parenthesis containing a single condition, we should just have:

```
                        has_capability('mod/coursework:addagreedgrade', $context)
```

opitz replied:
removed.

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+                $submission = submission::find($sub);
+                if ($coursework->assessor_has_any_allocation_for_student($submission->reload()->get_allocatable()) ||
+                    (has_capability('mod/coursework:addagreedgrade', $context))
+                    && !empty($submission)
```
The first condition is <code>$coursework->assessor_has_any_allocation_for_student(<b>$submission</b>->reload()->get_allocatable()</code> so surely either `!empty($submission)` is implicitly true by this point or there'll be an error before we get this far?

opitz replied:
refactored

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+                            && !$submission->get_coursework()->sampling_enabled())
+                        || ($submission->get_coursework()->sampling_enabled()
+                            && $submission->all_inital_graded()
+                            && $submission->max_number_of_feedbacks() > 1))) {
```
I think this is `if (a || b && !c && ((d && !e) || (f && g && h))) {`. Mixing `||` and `&&` in `a || b && ...` is tricky to read, is it correct? Could we add parenthesis to make it clear whether this means `(a || b) && ...` or `a || (b && ...)`? (I think PHP's operator precedence makes it the latter).

opitz replied:
refactored

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+        if ($canfinalgrade) {
+            $total = count($submissions);
+            $remaining = self::remove_final_gradable_submissions($submissions);
+            $needsgrading += ($total - count($remaining));
```
Surely this is short for:

```
            $needsgrading = 0 + ($total - count($remaining));
```

Instead we could just have:

```
            $needsgrading = $total - count($remaining);
```

or:

```
            $needsgrading = count($submissions) - count($remaining);
```

opitz replied:
refactored... :-)

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+                    || $submission->is_assessor_initial_grader()
+                    && (!has_capability('mod/coursework:administergrades',
+                            $submission->get_coursework()->get_context())
+                        && !is_siteadmin($USER->id))) {
```
Mix of `||` and `&&` without parenthesis again, i.e. `if (a || b && cc && d) {`. I think we should add brackets so that one doesn't have to look up operator precedence.

opitz replied:
refactored

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+        foreach ($submissions as $sub) {
+            $submission = submission::find($sub);
+
+            if (!empty($submission->get_final_grade() )) {
```
Rogue space: <code>if (!empty($submission->get_final_grade(<b>) )</b>) {</code>.

opitz replied:
removed

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,233 @@
+        // Remove unwanted submissions.
+        $submissions = self::remove_unfinalised_submissions($submissions);
+        $submissions = self::remove_ungradable_submissions($submissions);
+        $submissions = self::removed_final_graded_submissions($submissions);
```
The call to `get_submissions_for_assessor()` above loops through all submissions, and now we're going to loop through them three further times. Can't we have a single `remove_ungradable_submissions()` function which does something like:

```
        foreach ($submissions as $sub) {
            $submission = submission::find($sub);

            if (empty($submission->finalised)
                    || !empty($submission->get_final_grade())
                    || (has_capability('mod/coursework:addallocatedagreedgrade', $submission->get_coursework()->get_context())
                    && !$submission->is_assessor_initial_grader() && $submission->all_inital_graded())) {
                unset($submissions[$sub->id]);
            }
        }
        return $submissions;
```

opitz replied:
Thanks, refactored.

opitz committed:
addressed code review feedback

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -40,50 +40,57 @@ class submission_figures {
-                if ( $submission->final_grade_agreed()) {
+                if ($submission->final_grade_agreed()) {
+                    $submission->submissiondatetime = $submission->timesubmitted; // We need that for the feedback tracker.
                     continue;
```
Why `continue`? There's no code to be skipped in the current loop.

opitz replied:
removed.

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -106,9 +113,7 @@ public static function calculate_needsgrading_for_assessor($coursework): int {
         $submissions = self::get_submissions_for_assessor($coursework);
```
```
103         $needsgrading = 0;
            ⋮
106         if (!$coursework->has_multiple_markers()
107                 && !$coursework->allocation_enabled()
108                 && !has_capability('mod/coursework:addinitialgrade', $context)
109                 && has_capability('mod/coursework:addagreedgrade', $context)) {
110             return $needsgrading;
111         }
```

I think if we instead have `return 0` it would be clearer what we're doing, e.g.:

```
        $context = $coursework->get_context();

        if (!$coursework->has_multiple_markers()
                && !$coursework->allocation_enabled()
                && !has_capability('mod/coursework:addinitialgrade', $context)
                && has_capability('mod/coursework:addagreedgrade', $context)) {
            return 0;
        }

        $needsgrading = 0;
```

opitz replied:
I agree, refactored.

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -218,13 +190,13 @@ private static function get_assessor_initial_graded_submissions(array $submissio
+            $hasallfeedbacks = count($submission->get_assessor_feedbacks()) >= $submission->max_number_of_feedbacks();
+            $isinitialgrader = $submission->is_assessor_initial_grader();
+            $context = $submission->get_coursework()->get_context();
+            $isadmin = has_capability('mod/coursework:administergrades', $context) || is_siteadmin($USER->id);
```
```
195             $context = $submission->get_coursework()->get_context();
196             $isadmin = has_capability('mod/coursework:administergrades', $context) || is_siteadmin($USER->id);
```

`$isadmin` (and `$context`) is constant within this function so we don't need to repeatedly determine this within the `foreach`.

opitz replied:
I probably don't get it: $context is a function of $submission which is determined inside the loop, so I think it needs to stay there?

leonstr replied:
`get_assessor_initial_graded_submissions()` is called with submissions for a single mod_coursework instance. Whether `$isadmin` is true or not is a property of that instance not its submissions. Thus `$isadmin` is constant across all submissions being iterated through here so should not be determined in each iteration.

So, possibly:

```
    public static function calculate_needsgrading_for_assessor(int $instance): int {            
        global $USER;
        ⋮
            $isadmin = has_capability('mod/coursework:administergrades', $context) || is_siteadmin($USER->id);
            $needsgrading += count(self::get_assessor_initial_graded_submissions($submissions, $isadmin));
        ⋮
     }
    ⋮
    private static function get_assessor_initial_graded_submissions(array $submissions, bool $isadmin): array {
        foreach ($submissions as $submission) {
            $hasallfeedbacks = count($submission->get_assessor_feedbacks()) >= $submission->max_number_of_feedbacks();
            $isinitialgrader = $submission->is_assessor_initial_grader();

            // Remove submissions that are not assessable by the current user at this stage.
            if ($hasallfeedbacks || ($isinitialgrader && !$isadmin)) {
                unset($submissions[$submission->id]);
            }
        }
        return $submissions;
    }
```

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -218,13 +190,13 @@ private static function get_assessor_initial_graded_submissions(array $submissio
         foreach ($submissions as $sub) {
             $submission = submission::find($sub);
```
In `get_submissions_for_assessor()` we call `submission::find($sub)`, then in functions we pass the results to we again call:

```
152             $submission = submission::find($sub);
    ⋮
172             $submission = submission::find($sub);
    ⋮
191             $submission = submission::find($sub);
```

Provided we ensure the initial results contains `submission` instances – which I think just needs a change to the first `if` block:

```
 47         if (!$coursework->has_multiple_markers()
 48                 && has_capability('mod/coursework:addagreedgrade', $context)
 49                 && !has_capability('mod/coursework:addinitialgrade', $context)) {
 50             foreach ($submissions as $sub) {
 51                 $submission = submission::find($sub);
 52 
 53                 if (count($submission->get_assessor_feedbacks()) < $submission->max_number_of_feedbacks()) {
 54                     continue;
 55                 }
 56 
 57                 if ($submission->final_grade_agreed()) {
 58                     $submission->submissiondatetime = $submission->timesubmitted; // We need that for the feedback tracker.
 59                 }
 60                 $assessorsubmissions[] = $submission;
 61             }
 62         } else if (is_siteadmin($USER) || ...
```

– then presumably we don't need these subsequent calls to `submission::find()`? If so we'd also need to remove the unneeded variable name by changing `$sub` to `$submission` in the subsequent `foreach`s and `unset()`s:

```
151         foreach ($submissions as $submission) {
                ⋮
158                 unset($submissions[$submission->id]);
            ⋮
171         foreach ($submissions as $submission) {
                ⋮
175                 unset($submissions[$submission->id]);
            ⋮
190         foreach ($submissions as $submission) {
                ⋮
200                 unset($submissions[$submission->id]);
```

opitz replied:
Thank you very much! Refactored.

opitz committed:
addressed code review feedback 2

opitz committed:
CTP-4889: refactored get_submissions_for_assessor()

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
+        $canaddinitial = has_capability('mod/coursework:addinitialgrade', $context);
+        $isadmin = is_siteadmin($USER);
+        $allocationdisabled = !$coursework->allocation_enabled();
+        $canoverride = has_capability('mod/coursework:administergrades', $context);
```
We use `$isadmin` below:

```
 73             // Case 2: Admin or no allocation or grading override.
 74             if ($isadmin || $allocationdisabled || $canoverride) {
```

But `$canoverride` is the result of `has_capability('mod/coursework:administergrades', $context)`, and `has_capability()` will return true for a site admin. Do we need an additional check that they're a site admin? I don't think we do.

opitz replied:
gone now...

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
+                $assessorsubmissions[$submission->id] = $submission;
             }
         }
 
```
```
189             $isadmin = has_capability('mod/coursework:administergrades', $context) || is_siteadmin($USER->id);
```

`has_capability()` will return true for a site admin, so I think this just needs to be:

```
            $isadmin = has_capability('mod/coursework:administergrades', $context);
```

opitz replied:
cheers

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
-                }
-
                 $assessorsubmissions[] = $submission;
+                continue;
```
Would this not be better with if/else-if/else rather than jumping out the loop with `continue`? For example:

```
            // Case 1: Agreed grade only, and not all markers done yet.
            if ($singlemarker && $canaddagreed && !$canaddinitial) {
                if (count($submission->get_assessor_feedbacks()) === $submission->max_number_of_feedbacks()) {
                    $assessorsubmissions[] = $submission;
                }

            // Case 2: Admin or no allocation or grading override.
            } else if ($isadmin || $allocationdisabled || $canoverride) {
                $assessorsubmissions[$submission->id] = $submission;

            // Case 3: Allocated assessor or allowed to add agreed grade after initial grading (and sampling if enabled).
            } else {
                $allocated = $coursework->assessor_has_any_allocation_for_student($submission->reload()->get_allocatable());
                $sampled = $submission->get_coursework()->sampling_enabled();
                $initialgraded = $submission->all_inital_graded();
                $requiresamplingcheck = $sampled && ($submission->max_number_of_feedbacks() > 1);

                if ($allocated || ($canaddagreed && $initialgraded && (!$sampled || $requiresamplingcheck))) {
                    $assessorsubmissions[$submission->id] = $submission;
                }
            }
```

opitz replied:
Thank you very much, this is better...!

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
-                    $submission->submissiondatetime = $submission->timesubmitted; // We need that for the feedback tracker.
-                }
-
                 $assessorsubmissions[] = $submission;
```
Is there a reason why this doesn't have an array key, whereas for other conditions we have:

```
 75                 $assessorsubmissions[$submission->id] = $submission;
```

and

```
 86                 $assessorsubmissions[$submission->id] = $submission;
```

opitz replied:
The reason must have been my distracted brain... thanks!

opitz committed:
addressed code review feedback 3

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
         global $USER;
```
`$USER` not used in function.

opitz replied:
removed.

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -42,54 +42,48 @@ public static function get_submissions_for_assessor(int $instance): array {
+                $assessorsubmissions[$submission->id] = $submission;
             }
         }
 
```
```
179         global $USER;
```

`$USER` not used in function.

opitz replied:
removed.

leonstr committed:
Attempt to fix unexpected Behat failures

Behat tests have been unexpectedly failing. These failures were with
Chrome 138.0.7204.100. Switching to Firefox to see if that makes a
difference.

leonstr committed:
Fix Behat error Invalid profile passed

Previous attempt to run Behat tests failed with "Invalid profile
passed: headlessfirefox".

leonstr commented on classes/services/submission_figures.php:
```diff
@@ -45,15 +45,14 @@ public static function get_submissions_for_assessor(int $instance): array {
         $context = $coursework->get_context();
```
We have functions `remove_ungradable_submissions()`, `remove_final_gradable_submissions()` and `get_assessor_initial_graded_submissions()` both here and in `renderers/object_renderer.php` (with the same signature).

`remove_ungradable_submissions()` looks slightly different, but is it functionally different? If it is different, should it have a different name? Otherwise could we make the existing function public and call that instead?

`remove_final_gradable_submissions()` and `get_assessor_initial_graded_submissions()` appear to be identical to those in `renderers/object_renderer.php` could we not just call those instead?

opitz replied:
They are indeed (functionally) identical. When I wrote the new functions to return the assessor related data I copied all functions required as well because I did not want to interfere with your possible changes to that part of code.
So if you think that the original functions in object_renderer.php will survive your refactoring they surely may be called.

leonstr reviewed (commented):
Switching to `--profile firefox` in `moodle-ci.yml` appears to have resolved the unexpected Behat test failures (now: _143 scenarios (143 passed)_).

opitz committed:
addressed code review feedback

watson8 commented on classes/framework/table_base.php:
```diff
@@ -57,6 +57,26 @@ abstract class table_base {
+    /**
+     * @var int
+     */
+    public $timecreated;
```
Most or all of classes in models extend table_base and they all have different properties so I think we should avoid adding these instance vars to the base class.  I assume that this issue relates to the moderator class? The way we handled this in the coursework class was to add #[\AllowDynamicProperties] to the coursework class

opitz replied:
removed again.

watson8 commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,192 @@
+ * Class to provide figures around submissions and gradings.
+ */
+class submission_figures {
+
```
Having this class is a good idea.  While at present there is code duplication with the code in object_renderer (used to calculate the stats for the marking summary box), that other code is horrible and needs refactoring.  So I think the way to go would be to keep this code here for now, and have another ticket to remove all the code in object_renderer, and instead render the marking summary table from mustache, using counts produced from this class.  I.e. this class can be expanded to include the additional counts that the marking summary template will need
<img width="349" height="256" alt="image" src="https://github.com/user-attachments/assets/9c1eecc8-2710-454a-825d-8a1957d3a127" />

opitz replied:
@watson8  I am not sure what you want me to do here..?

watson8 replied:
I don't think anything needs doing for it on this PR but opened a new ticket for the backlog https://ucldata.atlassian.net/browse/CTP-5288

watson8 reviewed (commented):
I left a couple of comments only one of which I think needs action. This branch would benefit from being rebased

opitz committed:
allowing dynamic properties for class table_base

opitz committed:
Addressed Copilot code review issue: renamed all_inital_graded() to all initial_graded()

opitz commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,192 @@
+     */
+    private static function remove_final_gradable_submissions(array $submissions): array {
+        foreach ($submissions as $submission) {
+            if (!empty($submission->all_inital_graded()) ) {
```
refactored

opitz commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,192 @@
+            if (empty($submission->finalised)
+                || !empty($submission->get_final_grade())
+                || (has_capability('mod/coursework:addallocatedagreedgrade', $submission->get_coursework()->get_context())
+                    && !$submission->is_assessor_initial_grader() && $submission->all_inital_graded())) {
```
refactored

opitz commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,192 @@
+     */
+    private static function remove_final_gradable_submissions(array $submissions): array {
+        foreach ($submissions as $submission) {
+            if (!empty($submission->all_inital_graded()) ) {
```
refactored

opitz commented on classes/services/submission_figures.php:
```diff
@@ -0,0 +1,192 @@
+            } else {
+                $allocated = $coursework->assessor_has_any_allocation_for_student($submission->reload()->get_allocatable());
+                $sampled = $submission->get_coursework()->sampling_enabled();
+                $initialgraded = $submission->all_inital_graded();
```
refactored

opitz committed:
add explanation for allowing dynamic properties
